basicstation, sx1302_hal: initial commit

mars642 opened the pull request:
Maintainer: me
Compile tested: x86_64, lantiq_xrx200, latest openwrt-21.02, latest openwrt-22.03
Run tested: x86_64, lanitq_xrx200, latest openwrt-21.02, latest openwrt-22.03

Description:
(LoRa) Basicstation is an implementation of a LoRa packet forwarder and is intended to be run on the host of a LoRa-based gateway. Basicstation forwards RF packets recieved by a concentrator to a LoRaWAN network server (LNS). It also transmits RF packets received from the LNS to one or multiple LoRa end devices.
Further information: https://lora-developers.semtech.com/build/software/lora-basics/lora-basics-for-gateways

The SX1302 hardware abstraction layer is required by the basicstation to control the concentrator based on SX1302/SX1303 chip.

Testing:
Several Class-A end devices from different manufactures were used. TTN and ChirpStack were tested as LNS servers. The concentrator card used is based on Semtech's SX1302/SX1303 corecell reference design and does not have GPS. Currently only concentrator cards with USB interface are supported. (no SPI support)

Inspired by lora-feed from @xueliu 

Signed-off-by: Marcus Schref mschref@tdt.de

mars642 commented:
ping @aparcar

aparcar commented:
Looks excellent to me, amazing first time contribution!

I don't have hardware at hand to test this, do you have a recommendation? That's not a blocker for this PR but I'm keen to test it on real hardware.

mars642 commented:
Thank you for your feedback!

I have selected the n-fuse LRWCC3-MPCI concentrator card for testing. It is based on the SX1303 chip from Semtech and most importantly, it is in stock right now.
Here is a link: https://www.n-fuse.co/de/devices/SX1303-LoRaWAN-Concentrator-Card-mini-PCIe.html
Make sure to select the USB version.

In my opinion, you can choose any end-devices, as Class-A support is mandatory.

BKPepe commented on libs/sx1302_hal/Makefile:
```diff
@@ -0,0 +1,120 @@
+#
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=sx1302_hal
+PKG_VERSION:=2.1.0
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/Lora-net/sx1302_hal.git
+PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_VERSION)
+PKG_SOURCE_VERSION:=4b42025d1751e04632c0b04160e0d29dbbb222a5
+PKG_SOURCE_DATE:=2021-04-22
+PKG_MIRROR_HASH:=8733a2cf594555c17529c682f79612d178a7f473afdaf962c7bfd65555363b63
```
This does not look good to me. In the first place, you specified the exact version, but then you used ``PKG_SOURCE_DATE``, ``PKG_SOURCE_VERSION``.

No need to check out sources from Git, if they are using tags and releases. Refer to other Makefiles, which they are using e.g. codeload

BKPepe commented on libs/sx1302_hal/Makefile:
```diff
@@ -0,0 +1,120 @@
+PKG_SOURCE_DATE:=2021-04-22
+PKG_MIRROR_HASH:=8733a2cf594555c17529c682f79612d178a7f473afdaf962c7bfd65555363b63
+
+PKG_MAINTAINER:= Marcus Schref <mschref@tdt.de>
+
+PKG_LICENSE_FILES:=LICENSE.TXT
```
```suggestion
PKG_MAINTAINER:=Marcus Schref <mschref@tdt.de>
PKG_LICENSE_FILES:=LICENSE.TXT
```

By looking at https://github.com/Lora-net/sx1302_hal/blob/master/LICENSE.TXT , I would say that the LICENSE is MIT and BDS-3-Clause.

BKPepe commented on libs/sx1302_hal/Makefile:
```diff
@@ -0,0 +1,120 @@
+endef
+
+define Package/sx1302_hal/description
+SX1302/SX1303 Hardware Abstraction Layer
```
```suggestion
  SX1302/SX1303 Hardware Abstraction Layer
```

BKPepe commented on libs/sx1302_hal/Makefile:
```diff
@@ -0,0 +1,120 @@
+SX1302/SX1303 Hardware Abstraction Layer
+endef
+
+define Package/sx1302_hal-tests
+	SECTION:=net
+	CATEGORY:=Network
+	SUBMENU:=LoRaWAN
+	TITLE:=Tests for SX1302/SX1303 HAL
+	DEPENDS:=+sx1302_hal +kmod-usb-acm
+endef
+
+define Package/sx1302_hal-utils
+	SECTION:=net
+	CATEGORY:=Network
+	SUBMENU:=LoRaWAN
+	TITLE:=Utilities for SX1302/SX1303 HAL
+	DEPENDS:=+sx1302_hal +kmod-usb-acm
+endef
```
Inconsistency. In the first define - ``Package/sx1302_hal``, you have two spaces as indentation and here tab. Stick to two spaces.

```suggestion
define Package/sx1302_hal-tests
  SECTION:=net
  CATEGORY:=Network
  SUBMENU:=LoRaWAN
  TITLE:=Tests for SX1302/SX1303 HAL
  DEPENDS:=+sx1302_hal +kmod-usb-acm
endef

define Package/sx1302_hal-utils
  SECTION:=net
  CATEGORY:=Network
  SUBMENU:=LoRaWAN
  TITLE:=Utilities for SX1302/SX1303 HAL
  DEPENDS:=+sx1302_hal +kmod-usb-acm
endef
```

BKPepe commented on libs/sx1302_hal/patches/000-edit_cflags.patch:
```diff
@@ -0,0 +1,22 @@
+--- a/libloragw/Makefile
++++ b/libloragw/Makefile
```
Missing Git header, which contains usual information like: who is the author, the commit subject, commit description and from that, we can see if this is a hack or it could be upstreamed.

BKPepe commented on libs/sx1302_hal/patches/001-add_compatibility_with_basicstation.patch:
```diff
@@ -0,0 +1,10 @@
+--- a/libloragw/inc/loragw_hal.h
++++ b/libloragw/inc/loragw_hal.h
```
Missing Git header, which contains usual information like: who is the author, the commit subject, commit description and from that, we can see if this is a hack or it could be upstreamed.

BKPepe commented on net/basicstation/Makefile:
```diff
@@ -0,0 +1,66 @@
+PKG_LICENSE:=BSD-3-Clause
+PKG_LICENSE_FILES:=LICENSE
+
+PKG_BUILD_DEPENDS:=sx1302_hal
+PKG_BUILD_DEPENDS+=mbedtls
```
You don't need to define twice PKG_BUILD_DEPENDS.

BKPepe commented on net/basicstation/Makefile:
```diff
@@ -0,0 +1,66 @@
+PKG_VERSION:=2.0.6
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/lorabasics/basicstation
+PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_VERSION)
+PKG_SOURCE_VERSION:=ba4f85d80a438a5c2b659e568cd2d0f0de08e5a7
+PKG_SOURCE_DATE:=2022-01-18
+PKG_MIRROR_HASH:=d42ed480740d6cf9db0ed072626978fe88df5b42d9e19dbc447d83ff801b4d46
```
Same as above

BKPepe commented on net/basicstation/Makefile:
```diff
@@ -0,0 +1,66 @@
+PKG_SOURCE_DATE:=2022-01-18
+PKG_MIRROR_HASH:=d42ed480740d6cf9db0ed072626978fe88df5b42d9e19dbc447d83ff801b4d46
+
+PKG_MAINTAINER:= Marcus Schref <mschref@tdt.de>
+
+PKG_LICENSE:=BSD-3-Clause
+PKG_LICENSE_FILES:=LICENSE
```
```suggestion
PKG_MAINTAINER:=Marcus Schref <mschref@tdt.de>
PKG_LICENSE:=BSD-3-Clause
PKG_LICENSE_FILES:=LICENSE
```

BKPepe commented on net/basicstation/patches/000-include_sys_time.patch:
```diff
@@ -0,0 +1,11 @@
+--- a/src/aio.c
```
Did you try to upstream this patch?

mars642 replied:
Thanks for your review! I have opened a pull request for this patch: https://github.com/lorabasics/basicstation/pull/165

BKPepe commented on net/basicstation/Makefile:
```diff
@@ -0,0 +1,66 @@
+endef
+
+define Package/basicstation/description
+LoRa Basic Station. The LoRaWAN Gateway Software.
```
```suggestion
  LoRa Basic Station. The LoRaWAN Gateway Software.
```

BKPepe reviewed (changes requested):
I

BKPepe commented on net/basicstation/Makefile:
```diff
@@ -0,0 +1,66 @@
+
+PKG_NAME:=basicstation
+PKG_VERSION:=2.0.6
+PKG_RELEASE:=1
```
As this is inspired from https://github.com/xueliu/lora-feed/commits/master/basicstation, why don't you use autorelease?

mars642 replied:
I missed that, but now it is updated.

Hurricos commented:
> Currently only concentrator cards with USB interface are supported. (no SPI support)

Hi there,

I have an older MTCDT with an SPI-based SX1302 in it which I'm working on standing OpenWrt onto. Is there a reason you're not providing SPI support in this package -- is it a matter of basicstation configuration?

mars642 commented:
Hi @Hurricos,

at the moment i don't have suitable hardware to test SPI-based devices. Maybe you could add SPI support in future versions since you have the necessary hardware. If so, I could review your contribution.

BKPepe commented on libs/sx1302_hal/Makefile:
```diff
@@ -0,0 +1,116 @@
+  SECTION:=net
+  CATEGORY:=Network
+  SUBMENU:=LoRaWAN
+  TITLE:=Tests for SX1302/SX1303 HAL
```
Shouldn't be those test in one category? Right now, it is in two categories in ``make menuconfig``. What about something like this or vice versa?

```
define Package/sx1302_hal-tests
  $(call Package/sx1302_hal)
  TITLE:+=(Tests)
  DEPENDS:=sx1302_hal +kmod-usb-acm
endef
```

Also, there is a missing submenu for the first package.

mars642 replied:
You are right. All three packages are now located at -> Libraries -> IoT.

BKPepe commented on libs/sx1302_hal/patches/000-edit_cflags.patch:
```diff
@@ -0,0 +1,42 @@
+Date: Mon, 22 Aug 2022 10:02:13 +0200
+Subject: [PATCH] sx1302_hal: edit cflags
+
+Enable use of TARGET_CFLAGS in addition to specified CFLAGS
```
Did you try to upstream this? Otherwise, we will need to carry this patch forever.

mars642 replied:
I opened pull requests for both sx1302_hal patches.
https://github.com/Lora-net/sx1302_hal/pull/85
https://github.com/Lora-net/sx1302_hal/pull/86

BKPepe commented on libs/sx1302_hal/patches/001-add_compatibility_with_basicstation.patch:
```diff
@@ -0,0 +1,27 @@
+Date: Mon, 22 Aug 2022 10:06:03 +0200
+Subject: [PATCH] sx1302_hal: add compatibility with basicstation
+
+Add LGW_LBT_ISSUE define used by Basicstation
```
Did you try to upstream this? Otherwise, we will need to carry this patch forever.

BKPepe commented on net/basicstation/Makefile:
```diff
@@ -0,0 +1,60 @@
+PKG_SOURCE_URL:=https://codeload.github.com/lorabasics/basicstation/tar.gz/v$(PKG_VERSION)?
+PKG_HASH:=7e285de94bae1174b4c3496fc3ab15fe67c72f92c0693d2320bafc654a9dfb43
+
+PKG_MAINTAINER:= Marcus Schref <mschref@tdt.de>
```
```suggestion
PKG_MAINTAINER:=Marcus Schref <mschref@tdt.de>
```

BKPepe commented on net/basicstation/Makefile:
```diff
@@ -0,0 +1,60 @@
+
+define Build/Prepare
+	$(call Build/Prepare/Default)
+	rm -rf $(PKG_BUILD_DIR)/deps/
+	rm -f $(PKG_BUILD_DIR)/makefile
+	rm -f $(PKG_BUILD_DIR)/makefile.s2core
+	rm -f $(PKG_BUILD_DIR)/setup.gmk
```
Default prepare does not work for this package?

mars642 replied:
By default, Basicstation would pull dependencies (mbedtls and sx1302_hal) via some custom scripts. To avoid this behavior and use OpenWrt packages, i removed the dependencies folder and the Makefiles and added CMake. This topic was also mentioned in https://github.com/xueliu/lora-feed/issues/27.

BKPepe commented on libs/sx1302_hal/Makefile:
```diff
@@ -0,0 +1,116 @@
+
+define Package/sx1302_hal-utils/install
+	$(INSTALL_DIR) $(1)/usr/bin
+
```
Just nitpick. No need to have here empty row

```suggestion
```

BKPepe reviewed (commented):
What about adding ``test.file`` to be run tested inside CI to check versioning and if it works?

mars642 commented:
> What about adding `test.file` to be run tested inside CI to check versioning and if it works?

I saw the test.sh file you authored at https://github.com/openwrt/packages/commit/21992344b54144dae1b73ad7145540328e7ed0bb and adopted it to this package. Is it OK like that?

BKPepe commented:
Seems good for basicstation. See output in test build:
```
Configuring basicstation.
Use package specific test.sh
Station: 2.0.6 
Test successful
Removing package basicstation from root...
Failed to connect to ubus
Not deleting modified conffile /etc/config/basicstation.
Testing package sx1302_hal-tests in version 2.1.0 from sx1302_hal
Unknown package 'sx1302_hal-tests'.
Collected errors:
 * pkg_hash_check_unresolved: cannot find dependency sx1302_hal for sx1302_hal-tests
 * pkg_hash_fetch_best_installation_candidate: Packages for sx1302_hal-tests found, but incompatible with the architectures configured
 * opkg_install_cmd: Cannot install package sx1302_hal-tests.
No test.sh script available
No packages removed.
Testing package sx1302_hal-utils in version 2.1.0 from sx1302_hal
Unknown package 'sx1302_hal-utils'.
Collected errors:
 * pkg_hash_check_unresolved: cannot find dependency sx1302_hal for sx1302_hal-utils
 * pkg_hash_fetch_best_installation_candidate: Packages for sx1302_hal-utils found, but incompatible with the architectures configured
 * opkg_install_cmd: Cannot install package sx1302_hal-utils.
No test.sh script available
No packages removed.
Testing package sx1302_hal in version 2.1.0 from sx1302_hal
Installing sx1302_hal (2.1.0-1) to root...
Configuring sx1302_hal.
No test.sh script available
Removing package sx1302_hal from root...
```

mars642 commented:
> Seems good for basicstation. See output in test build:
> 
> ```
> Configuring basicstation.
> Use package specific test.sh
> Station: 2.0.6 
> Test successful
> Removing package basicstation from root...
> Failed to connect to ubus
> Not deleting modified conffile /etc/config/basicstation.
> Testing package sx1302_hal-tests in version 2.1.0 from sx1302_hal
> Unknown package 'sx1302_hal-tests'.
> Collected errors:
>  * pkg_hash_check_unresolved: cannot find dependency sx1302_hal for sx1302_hal-tests
>  * pkg_hash_fetch_best_installation_candidate: Packages for sx1302_hal-tests found, but incompatible with the architectures configured
>  * opkg_install_cmd: Cannot install package sx1302_hal-tests.
> No test.sh script available
> No packages removed.
> Testing package sx1302_hal-utils in version 2.1.0 from sx1302_hal
> Unknown package 'sx1302_hal-utils'.
> Collected errors:
>  * pkg_hash_check_unresolved: cannot find dependency sx1302_hal for sx1302_hal-utils
>  * pkg_hash_fetch_best_installation_candidate: Packages for sx1302_hal-utils found, but incompatible with the architectures configured
>  * opkg_install_cmd: Cannot install package sx1302_hal-utils.
> No test.sh script available
> No packages removed.
> Testing package sx1302_hal in version 2.1.0 from sx1302_hal
> Installing sx1302_hal (2.1.0-1) to root...
> Configuring sx1302_hal.
> No test.sh script available
> Removing package sx1302_hal from root...
> ```

I added a test file for sx1302_hal packages and fixed the errors.
![grafik](https://user-images.githubusercontent.com/107621362/187859189-75b23362-50ec-4a0f-bf8e-5a72798ff674.png)

BKPepe commented on net/basicstation/Makefile:
```diff
@@ -0,0 +1,60 @@
+	$(INSTALL_DIR) $(1)/usr/bin
+	$(INSTALL_BIN) $(PKG_BUILD_DIR)/station $(1)/usr/bin
+	$(INSTALL_DIR) $(1)/etc/config
+	$(INSTALL_BIN) ./files/etc/config/basicstation $(1)/etc/config/basicstation
```
```suggestion
	$(INSTALL_CONF) ./files/etc/config/basicstation $(1)/etc/config/basicstation
```

According to https://openwrt.org/docs/guide-developer/packages#file_installation_macros

BKPepe reviewed (approved):
Apart from one thing, Makefiles looks good! Test files were added. Much appreciated. Patches were submitted to upstream repositories for review. I am satisfied. :blush:

mars642 commented:
Glad to hear that! I replaced the installation macro, so it should be fine now.

BKPepe reviewed (approved):
One small thing was addressed and my last comment still stands. :)

feckert commented:
Well then let's try it
